### ISSUE TITLE
Error code 400 for invalid request

### ISSUE THREAD

**Issue by bitcity** [pennersr/django-allauth#1229]
I'm validating a sign-up form with an AJAX request. When the form is invalid (e.g. username is taken), the status code is `400` which is [defined as](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html) : 

> The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.

I checked the username creation part for gmail, twitter & a few other services. The network request returns a status code `200` even when the username is unavailable. 

I'm raising a request to re-consider using a status code `200` for cases like above.

**Comment by pennersr** [comment]
Returning 400 is actually quite common for AJAX calls. It comes in especially handy if you are using promises for your calls: `.then(function() { ..ok..}, function() { ..error..})` -- otherwise you would end up having to if in your success response to see if it really is a successful call you made.

As to what 4xx status code to return, opinions vary... though 400 is not a strange choice:

- http://stackoverflow.com/questions/6123425/rest-response-code-for-invalid-data
- http://www.bennadel.com/blog/2434-http-status-codes-for-invalid-data-400-vs-422.htm
- In line with e.g. the behavior of django rest framework.

All in all, I prefer to keep it this way ...

**Comment by bitcity** [comment]
Fair enough. I appreciate your justification & efforts.

**Comment by astarrh** [comment]
I'd like to second this request. To me as a Django user, `400` means there is something wrong with my code. I spent a while trying to figure out what could be wrong with my urls or POST request when in reality, I was just submitting incorrect information.

Django's standard practice is to return `200` if the request was received, along with the errors if there are any. If the request is successful, we can check for errors in the promise and handle them from there. It's just a little confusing for those of us coming from Django and still learning about Ajax. I had to dig around in the source files to finally figure this one out.

Thanks for supporting!
If this isn't going to change, at least a mention of this in the docs would be nice. Nothing came up for me when I searched them.
